nengo_loihi/simulator.py

```diff
@@ -184,9 +184,17 @@ def __init__(self, network, dt=0.001, seed=None, model=None,  # noqa: C901
             if len(self.host_pre.all_objects) > 0:
                 self.sims["host_pre"] = nengo.Simulator(
                     self.host_pre, dt=self.dt, progress_bar=False)
+
             if len(self.host.all_objects) > 0:
                 self.sims["host"] = nengo.Simulator(
                     self.host, dt=self.dt, progress_bar=False)
+            elif not precompute:
+                # If there is no host and precompute=False, then all objects
+                # must be on the chip, which is precomputable in the sense that
+                # no communication has to happen with the host.
+                # We could warn about this, but we want to avoid people having
+                # to specify `precompute` unless they absolutely have to.
+                self.precompute = True
```

Comment:
I don't like this. If I say `precompute=False`, I want that to be followed. Right now, there are all kinds of bugs in `precompute=True`. For example, it doesn't work well with large networks, in my experience as well as that of others. Though I think that might have to do with spike generation, so maybe it's a non-issue for networks without a host sim. Anyway, I think this change might be good long-term, when we've really sorted things out with `precompute`, but right now I'd rather things be dumb and just do what I ask.

Another option would be to have `precompute=None` as the default, in which case I would be fine with something like this, since then I can always say `precompute=False` explicitly and force it to be false.

Reply:
Like you say, I think the `precompute=True` bugs that we've encountered are irrelevant here because this case only occurs if there is nothing running on the host, meaning that we are not generating any spikes to inject into the model, we are simply letting Loihi run on its own with no interaction from us, then reading the values off the chip later. It seems silly to slow down the chip for no reason other than that we're nervous there might be bugs.

Speed argument aside, the main reasons I made this change is that it was a silly amount of work to get rid of the complexity warning for `run_steps`, requiring making a `_make_run_steps` method that then calls either `_make_emu_run_steps` and `_make_loihi_run_steps`, which both just barely pass the complexity check. If we allow `precompute=False` unequivocally, then we have another case in both of those methods, so they will fail the complexity check and we have to figure out another way to break up these methods. So while I agree that passing `precompute=None` and then determining it if not specified would be better, it mean that the `precompute=False` and no host case is possible, so we have to implement it regardless.

I don't disagree that explicitly setting `precompute=False` and having it change on you would be annoying, but I really can't imagine anyone being extremely opinionated about making sure their model is not precomputed while also having no nodes in their model. So, I guess I would be okay with setting `precompute=None` and then raising a `NotImplementedError` for having `precompute=False` and no `host`, because that would be a use case I think we would want to know about.

Reply:
OK, I'm fine keeping this for now. I agree that it does seem to be kind of silly just to avoid it because there could be bugs. If we do encounter any bugs, then we can revisit this. I just wish we knew more about why we're having issues with `precompute=True` on large models. I think the working hypothesis that it's due to the number of input spikes is still the best, but I don't think anybody has looked at this in detail.

Reply:
My hope is that we can speed up `precompute=False` and get rid of the precomputing logic before we have to look into it in detail :crossed_fingers: